### DISCUSSION TITLE
Make referencing nodes from script a one-click operation.

### DISCUSSION THREAD

**Discussion by ultramarinebicycle** [godotengine/godot-proposals#4078] [Editor]
My current setup to get node references:

1. Declare a NodePath export variable. 
2. Declare a Node variable.
3. Set the NodePath in the editor.
4. Set the Node according to the NodePath in ready().

Nodes are referenced very frequently through code and making this operation one-click will eliminate a few minutes off pretty much every scene in every game.

If you have a better way of doing this, let me know!

**Comment by mieldepoche** [comment]
There is a proposal for this here: #1048.
Right now you can do a one-liner but its very ugly: ```export(NodePath) onready var my_node = get_node(my_node) as Label```

**Reply by ultramarinebicycle** [reply]
Haha that's twisted indeed. Also as per the proposal, will the export variable remain correct should the NodePath change?

**Reply by Zireael07** [reply]
The suggested one-liner will need changing if you move the node.

**Reply by theraot** [reply]
By the way, with this code:

```
export(NodePath) onready var my_node = get_node(my_node) as Label
```

You can change the NodePath form the editor UI, no problem. The reason being that it would only be replaced with the Node at runtime. Well, unless we are talking about a tool script. If you want to support setting a NodePath in runtime, then you can use setget:

```
export(NodePath) var my_node setget set_my_node

func set_my_node(new_value) -> void:
	if new_value == null:
		my_node = null
		return

	if typeof(new_value) == TYPE_NODE_PATH:
		if not is_inside_tree():
			yield(self, "tree_entered")

		my_node = get_node(new_value) as Label
		return

	my_node = new_value as Label
```

By the way, Godot will call the setter during initialization to set the NodePath, so this also archives the same effect as the one liner, plus you can set a NodePath to it at any moment. *You will have an empty NodePath before ready.*

**Comment by yoskeleton** [comment]
This would be really hepful with big user interface, because we always rearrange everything ( i am never happy with my ui ) 

That the main reason i hate working with UI, 99% of my time is fixing nodepath everytime i move things

**Comment by theraot** [comment]
See also https://github.com/godotengine/godot-proposals/issues/4096

**Reply by ultramarinebicycle** [reply]
Did Juan see this thread 👀
